### PR TITLE
[KMCompiler][Ascend][NVIDIA] add is_strides_like_format operator

### PR DESCRIPTION
**Pull request by Lee050911**
## PR Category
Operator

## Type of Change
New Feature

## Description
This PR adds the `is_strides_like_format` operator, which checks whether the stride pattern of an input tensor matches the `channels_last` (4D) or `channels_last_3d` (5D) memory format. The operator is a lightweight host‑side metadata check with zero data copying.

- Supports only `"channels_last"` and `"channels_last_3d"` formats; all other format strings return `False`.
- Strict stride‑pattern comparison without relying on `torch.is_contiguous` (compatible with `torch.ops.aten.is_strides_like_format`).
- Implementations for both NVIDIA (CUDA) and Ascend (NPU) backends.

## Issue
N/A (new operator addition).

## Progress
- [x] Change is properly reviewed.
- [x] Change is fully covered by UT (fixed‑shape + 100 random‑shape tests).

### NVIDIA H20
Average latency in microseconds (0.003040 ms = 3.04 µs), speedup ≈ 1.0.

| Shape                | Format            | Torch (µs) | Gems (µs) | Speedup |
|----------------------|-------------------|------------|-----------|---------|
| (1073741824,)        | channels_last     | 3.04       | 3.04      | 1.00    |
| (1073741824,)        | channels_last_3d  | 3.04       | 3.04      | 1.00    |
| (64,64)              | channels_last     | 3.04       | 3.04      | 1.00    |
| (64,64)              | channels_last_3d  | 3.04       | 3.04      | 1.00    |
| (4096,4096)          | channels_last     | 3.04       | 3.04      | 1.00    |
| (4096,4096)          | channels_last_3d  | 3.04       | 3.04      | 1.00    |
| (64,512,512)         | channels_last     | 3.04       | 3.04      | 1.00    |
| (64,512,512)         | channels_last_3d  | 3.04       | 3.04      | 1.00    |
| (1024,1024,1024)     | channels_last     | 3.04       | 3.04      | 1.00    |
| (1024,1024,1024)     | channels_last_3d  | 3.04       | 3.04      | 1.00    |
| (2,3)                | channels_last     | 3.04       | 3.04      | 1.00    |
| (2,3)                | channels_last_3d  | 3.04       | 3.04      | 1.00    |
| (4,5,6)              | channels_last     | 3.04       | 3.04      | 1.00    |
| (4,5,6)              | channels_last_3d  | 3.04       | 3.04      | 1.00    |
| (2,3,4,5)            | channels_last     | 3.04       | 3.04      | 1.00    |
| (2,3,4,5)            | channels_last_3d  | 3.04       | 3.04      | 1.00    |
| (8,3,224,224)        | channels_last     | 3.04       | 3.04      | 1.00    |
| (8,3,224,224)        | channels_last_3d  | 3.04       | 3.04      | 1.00    |
| (2,3,4,5,6)          | channels_last     | 3.04       | 3.04      | 1.00    |
| (2,3,4,5,6)          | channels_last_3d  | 3.04       | 3.04      | 1.00    |

### Ascend NPU (910B)
All measurements at timer granularity (0.000160 ms = 0.16 µs), speedup ≈ 1.0.

| Shape                | Format            | Torch (µs) | Gems (µs) | Speedup |
|----------------------|-------------------|------------|-----------|---------|
| (1073741824,)        | channels_last     | 0.16       | 0.16      | 1.00    |
| (1073741824,)        | channels_last_3d  | 0.16       | 0.16      | 1.00    |
| (64,64)              | channels_last     | 0.16       | 0.16      | 1.00    |
| (64,64)              | channels_last_3d  | 0.16       | 0.16      | 1.00    |
| (4096,4096)          | channels_last     | 0.16       | 0.16      | 1.00    |
| (4096,4096)          | channels_last_3d  | 0.16       | 0.16      | 1.00    |
| (64,512,512)         | channels_last     | 0.16       | 0.16      | 1.00    |
| (64,512,512)         | channels_last_3d  | 0.16       | 0.16      | 1.00    |
| (1024,1024,1024)     | channels_last     | 0.16       | 0.16      | 1.00    |
| (1024,1024,1024)     | channels_last_3d  | 0.16       | 0.16      | 1.00    |
| (2,3)                | channels_last     | 0.16       | 0.16      | 1.00    |
| (2,3)                | channels_last_3d  | 0.16       | 0.16      | 1.00    |
| (4,5,6)              | channels_last     | 0.16       | 0.16      | 1.00    |
| (4,5,6)              | channels_last_3d  | 0.16       | 0.16      | 1.00    |
| (2,3,4,5)            | channels_last     | 0.16       | 0.16      | 1.00    |
| (2,3,4,5)            | channels_last_3d  | 0.16       | 0.16      | 1.00    |
| (8,3,224,224)        | channels_last     | 0.16       | 0.16      | 1.00    |
| (8,3,224,224)        | channels_last_3d  | 0.16       | 0.16      | 1.00    |
| (2,3,4,5,6)          | channels_last     | 0.16       | 0.16      | 1.00    |
| (2,3,4,5,6)          | channels_last_3d  | 0.16       | 0.16      | 1.00    |

All correctness tests (12 test cases including fixed‑shape and 100 random shapes) pass on both platforms.

## Files Changed
- `src/flag_gems/ops/is_strides_like_format.py`: Generic operator entry (NVIDIA backend).
- `src/flag_gems/runtime/backend/_ascend/ops/is_strides_like_format.py`: Ascend‑specific implementation.
- `src/flag_gems/ops/__init__.py`: Export operator.
- `src/flag_gems/runtime/backend/_ascend/ops/__init__.py`: Register Ascend backend.
- `src/flag_gems/__init__.py`: Import operator into package.
- `tests/test_is_strides_like_format.py`: Correctness & random tests.
- `benchmark/test_is_strides_like_format.py`: Performance benchmark.
- `conf/operators.yaml`: Register operator metadata.

## Notes
- On Ascend NPU, `torch.channels_last` conversion is not supported, but the stride‑checking logic works correctly (read‑only metadata). Both benchmark and correctness tests run and pass.
- All tests are device‑agnostic and validated on both NVIDIA H20 and Ascend 910B.

### CODE REVIEW

**Comment by huangyiqun** [comment]
Please note that the CI tests are failing.

File: benchmark/test_is_strides_like_format.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            time_us = timer.timeit(1000).median * 1e6
+            print(
+                f"Shape: {str(shape):<20} | Format: {fmt:<15} | Time: {time_us:.3f} us"
+            )
+
+
+if __name__ == "__main__":
+    main()
+def test_benchmark():
+    main()
```
Comment (bin913) [inline benchmark/test_is_strides_like_format.py]: Please add benchmark tests according to the FlagGems specification, and you can refer to the tests of existing operators in the repository

File: tests/test_is_strides_like_format.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            actual_cl.item() == expected_cl
+        ), f"Shape {shape} after conversion: expected {expected_cl}, got {actual_cl.item()}"
+
+
+# ---------- abnormal test ----------
+def test_unsupported_format():
+    x = torch.randn(2, 3)
+    with flag_gems.use_gems():
+        with pytest.raises(ValueError, match="Unsupported stride format"):
+            flag_gems.is_strides_like_format(x, "invalid_format")
```
Comment (bin913) [inline tests/test_is_strides_like_format.py]: Please add accuracy tests according to the FlagGems specification, and you can refer to the tests of existing operators in the repository

File: src/flag_gems/ops/is_strides_like_format.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    if format == "contiguous":
+        result = x.is_contiguous()
+    else:  # format == "channels_last"
+        if x.dim() != 4:
+            result = False
+        else:
+            result = x.is_contiguous(memory_format=torch.channels_last)
+
+    return torch.tensor(result, dtype=torch.bool, device=x.device)
```
Comment (bin913) [inline src/flag_gems/ops/is_strides_like_format.py]: FlagGems is a library based on Triton implementation, please do not use torch

File: src/flag_gems/runtime/backend/_ascend/ops/is_strides_like_format.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    if format == "contiguous":
+        result = x.is_contiguous()
+    else:  # format == "channels_last"
+        if x.dim() != 4:
+            result = False
+        else:
+            result = x.is_contiguous(memory_format=torch.channels_last)
+
+    return torch.tensor(result, dtype=torch.bool, device=x.device)
```
Comment (bin913) [inline src/flag_gems/runtime/backend/_ascend/ops/is_strides_like_format.py]: FlagGems is a library based on Triton implementation, please do not use torch

File: benchmark/test_is_strides_like_format.py (hunk 1 above)
Comment (Lee050911) [reply benchmark/test_is_strides_like_format.py]: Thanks for the review! The benchmark test has been rewritten following the FlagGems specification. I've replaced the custom timing logic with a proper class that inherits from base.Benchmark, similar to existing operators like test_is_all_true.py. The benchmark now correctly measures both Torch and Gems latency and reports speedup.

File: tests/test_is_strides_like_format.py (hunk 2 above)
Comment (Lee050911) [reply tests/test_is_strides_like_format.py]: Thanks for the feedback! The accuracy tests have been fully updated to follow the FlagGems specification. I've added:

Parametrized tests covering fixed shapes for both channels_last and channels_last_3d formats.

A random test with 100 random shapes to improve coverage.

A test for unsupported formats.
The tests now compare the operator's output directly with torch.ops.aten.is_strides_like_format.

File: src/flag_gems/ops/is_strides_like_format.py (hunk 3 above)
Comment (Lee050911) [reply src/flag_gems/ops/is_strides_like_format.py]: Thanks for pointing this out! I've removed all torch.is_contiguous calls from the implementation. The operator now performs strict stride‑pattern comparison using x.shape and x.stride() directly, without relying on any Torch API. This aligns with the FlagGems principle of avoiding Torch dependencies in operators.

File: src/flag_gems/runtime/backend/_ascend/ops/is_strides_like_format.py (hunk 4 above)
Comment (Lee050911) [reply src/flag_gems/runtime/backend/_ascend/ops/is_strides_like_format.py]: Thanks for the review! The Ascend backend implementation has been updated accordingly. It now performs the same strict stride‑pattern comparison without invoking torch.is_contiguous or any other Torch API. The implementation is self‑contained and uses only x.shape and x.stride().

**Comment by huangyiqun** [comment]
Thank you for your contribution. I have approve workflows to run.

**Comment by Lee050911** [comment]
@tengqm Hi, sorry to bother you. My PR #4460 is ready for review – all CI checks have passed and it's been pending for about a week. Could you please take a look when you have a moment? Thanks a lot! 🙏

**Comment by Caeruleann** [comment]
The diff is a bit noisy and contains some unrelated changes. Could you please clean up the worktree and keep this PR focused on the intended changes?

**Comment by Lee050911** [comment]
@Caeruleann  Thanks for the review! I've cleaned up the worktree and kept this PR focused on the intended changes. The Ascend CI environment is missing pandas, so I added a graceful skip for the benchmark. All correctness tests (12/12) have passed on both backends.

**Review by huangyiqun — DISMISSED** [review]
LGTM.

**Review by huangyiqun — COMMENTED** [review]
The code is already approved, but GitHub now reports conflicts with master. Please rebase before merge.

**Review by huangyiqun — COMMENTED** [review]
This was approved before the new test rule. The changed test still has use_gems calls, so the rebase also needs a public-API migration and fresh CI.

**Comment by huangyiqun** [comment]
Plz resolve conflicts.

**Comment by Lee050911** [comment]
@huangyiqun Hi, I've completed the rebase and the API migration, and squashed the commits into one. There are currently 3 workflows (rule-check, linter, unit-test) awaiting approval. Could you please approve them when you have a moment? Thanks!
